### PR TITLE
[GHF][mergebot] record ghstack dependencies in the commit message

### PR DESCRIPTION
**Pull request by izaitsevfb**
Currently all information about the dependencies of ghstack PRs (e.g. #105010) is stripped away:
https://github.com/pytorch/pytorch/blob/c984885809194e0a807b3f5543450fae4dfa841a/.github/scripts/trymerge.py#L1077-L1078

This PR adds this information back in a more compact form. All dependencies (PR numbers) of each PR in ghstack are recorded.

The resulting commit message will look like this (the last line is new):

> Mock title (#123)
> 
> Mock body text
> Pull Request resolved: https://github.com/pytorch/pytorch/pull/123
> Approved by: https://github.com/Approver1, https://github.com/Approver2
> ghstack dependencies: #1, #2

---

### Testing

Unit tests.

---

### Note Re: `# type: ignore[assignment]` in unit tests.

I did my due diligence to find alternatives. Unfortunately mypy [doesn't](https://github.com/python/mypy/issues/6713) support this [way of patching methods](https://docs.python.org/3/library/unittest.mock-examples.html#mock-patching-methods), and the alternatives are either extremely verbose or don't work for this case. I decided it's not worth the effort (since the problem is limited only to the unit test).

### CODE REVIEW

File: .github/scripts/test_trymerge.py [hunk 1]
```diff
@@ -806,5 +806,143 @@ def test_dont_ignore_flaky_failures(self, *args: Any) -> None:
...
 
 
+class TestGitHubPRGhstackDependencies(TestCase):
+    @mock.patch("trymerge.gh_get_pr_info", return_value=mock_gh_get_info())
+    def setUp(self, mock_gh_get_pr_info: mock.MagicMock) -> None:
+        self.pr = self.mock_pr(123, "Mock title", "Mock body text", closed=False)
+        pass
+
+    @staticmethod
+    def mock_pr(num: int, title: str, body: str, closed: bool) -> GitHubPR:
```
Comment (huydhn) [inline .github/scripts/test_trymerge.py:816]: I have a feeling that you don't need to devise your own mock here in `test_trymerge` unless there is a clear reason to do so.  The usual way to mock here is to use the generic patches below and uses actual PR in PyTorch:

```
@mock.patch("trymerge.get_rockset_results", side_effect=mocked_rockset_results)
@mock.patch("trymerge.gh_graphql", side_effect=mocked_gh_graphql)
```

Here is a recent example I have in mind https://github.com/pytorch/pytorch/pull/105098.  Before running the test, you can export `GITHUB_TOKEN` and `ROCKSET_API_KEY` readily.  When you run the test locally, `mocked_gh_graphql` would make a single query to PyTorch to get the PR info and cache them locally in `.github/scripts/gql_mocks.json`

**Review by huydhn — APPROVED** [review]
LGTM!

**Comment by ZainRizvi** [comment]
Hi @izaitsevfb, one of the tests added here depended on the referenced ghstack PR never closing. Since that PR is now merged, the github gql queries the tests makes now return different results, causing the test to fail.

I've disabled this test for now in #107385, can you please fix this test to be more resilient? Either mock parts of the PR or perhaps better yet create a new dummy github account, have it create the ghstack PR you want, and then discard that account's username/password :P

**Comment by izaitsevfb** [comment]
> Hi @izaitsevfb, one of the tests added here depended on the referenced ghstack PR never closing. Since that PR is now merged, the github gql queries the tests makes now return different results, causing the test to fail.

This is weird, I assumed all the requests are mocked.

cc @huydhn

**Comment by huydhn** [comment]
> This is weird, I assumed all the requests are mocked. I'll investigate.
> 
> cc @huydhn

I think so too.  I could run the test that Zain mentions fine locally.  So just curious if the file `gql_mocks.json` is removed and repopulated, that would explain the failure.  Basically, I could delete `gql_mocks.json` and run all the test again to repopulate `gql_mocks.json`.  However, the cached content will be the snapshot at that later time instead.  So it's indeed a good idea to prefer closed PRs as test samples so that their contents don't change.  Having a open test PR would also work, but we need to keep it open.

```
(py3.9) huydo@huydo-mbp scripts % pytest -v test_trymerge.py -k test_last_pushed_at
=================================================================== test session starts ====================================================================
platform darwin -- Python 3.9.17, pytest-7.2.0, pluggy-1.0.0 -- /Users/huydo/miniconda3/envs/py3.9/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/huydo/Storage/mine/pytorch/.github/scripts/.hypothesis/examples')
rootdir: /Users/huydo/Storage/mine/pytorch, configfile: pytest.ini
plugins: xdoctest-1.1.0, cpp-2.3.0, resume-0.0.1, rerunfailures-10.3, shard-0.1.2, flakefinder-1.1.0, hypothesis-6.56.4, xdist-3.0.2, profiling-1.7.0, repeat-0.9.1, anyio-3.6.2
collected 39 items / 38 deselected / 1 selected
Running 1 items in this shard: .github/scripts/test_trymerge.py::TestTryMerge::test_last_pushed_at

test_trymerge.py::TestTryMerge::test_last_pushed_at PASSED                                                                                           [100%]

============================================================= 1 passed, 38 deselected in 1.76s =============================================================
```

**Comment by huydhn** [comment]
Ah I get it now, as Zain clean up the `pushDate` field, the cache is invalidated as the input parameters have changed.  And this triggers the failures and the content is now indeed different

**Pull request closed without merging**